COR-1053: Index mandatory

DailyMats opened the pull request:
The way this library works is that it loops over a large amount of regular expressions, loaded from a .yaml file. The first expression to match will be used, meaning that the order of the expressions matters.

The change in this branch is the addition of a preprocessing layer that reduces the number of expressions to loop over, to save processing time. This way the way the expressions are matched does not change.

The idea for the preprocessing layer is that the expressions contain mandatory snippets of text that need to be present for the expression to match, and these snippets can be indexed in a tree structure that can be used to get the expressions that can possibly match for the input string in question.

Another important change in this branch is that it uses re2 (https://github.com/google/re2/) instead of boost::regex.

The API has not changed, so no changes are necessary for the user of the library.

sebmil-daily commented:
Very good !

cieplik commented on internal/SnippetIndex.cpp:
```diff
@@ -0,0 +1,184 @@
+  case '%':
+    // These are always snippet characters
+    return true;
+  default:
```
There's more: https://regex101.com/r/R7sknN/1 . But maybe it'll go away if we switch to `re2::Regexp`.

DailyMats replied:
Sure, I mostly looked at the ones that are often present in user-agent strings and that aren't regular expression operators.

cieplik commented on UaParser.cpp:
```diff
@@ -75,24 +94,42 @@ DeviceStore fill_device_store(const YAML::Node& device_parser) {
-      agent_store.regExpr.assign(value, boost::regex::optimize | boost::regex::normal);
+      agent_store.regExpr.assign(value);
+
+      for (auto e : AlternativeExpander::expandExpression(value)) {
```
Please switch to `const auto&`.

cieplik commented on UaParser.cpp:
```diff
@@ -115,36 +151,66 @@ struct UAStore {
+  }
+
+  ~UAStore() {
+    for (auto& e : deviceStore) {
```
Could the vectors hold `std::unique_ptr<T>`s instead?

DailyMats commented on UaParser.cpp:
```diff
@@ -75,24 +94,42 @@ DeviceStore fill_device_store(const YAML::Node& device_parser) {
-  if (regex_flag == true) {
-    device.regExpr.assign(device.regExpr.str(), boost::regex::optimize | boost::regex::icase | boost::regex::normal);
+
+  device.regExpr.assign(regex, !regex_flag);
```
Well, it's not "my" variable, so to speak. I tried to change as little as possible.

cieplik replied:
Yes, I know we inherited it :-). But now we're going to maintain and read this code (and it's local to the function).

DailyMats commented on UaParser.cpp:
```diff
@@ -75,24 +94,42 @@ DeviceStore fill_device_store(const YAML::Node& device_parser) {
+
+  device.regExpr.assign(regex, !regex_flag);
+
+  for (auto e : AlternativeExpander::expandExpression(regex)) {
```
Will change.

DailyMats commented on internal/SnippetIndex.h:
```diff
@@ -0,0 +1,46 @@
+class SnippetIndex {
+public:
+  typedef uint32_t SnippetId;
+  typedef std::set<SnippetId> SnippetSet;
```
No, they need to be sorted and unique. Or at least those were my initial constraints. Maybe it's not actually necessary.

DailyMats replied:
Actually, the main point is that I use `find()` on the set to check if a snippet is present. So no, it can't easily be replaced with an `std::vector`.

DailyMats commented on UaParserTest.cpp:
```diff
@@ -109,6 +112,37 @@ TEST(DeviceFamily, test_device) {
+  test_expand("(a|b)", {"(a)", "(b)"});
+  test_expand("(?:a|b)", {"(?:a)", "(?:b)"});
+  test_expand("(a|b)?", {"(a|b)?"});
+  test_expand("(a|b)+", {"(a)+", "(b)+"});
```
The regexes are not supposed to be equivalent. The original regex is the one that will be used in the actual matching.

The logic is that if snippet "a" is found, it's worth matching the regex.

Expanding the expressions with this class allows the `SnippetIndex` to remain simpler (it ignores `(a|b)`, since neither is mandatory).

cieplik replied:
Thanks, that clarifies things.

DailyMats commented on internal/AlternativeExpander.h:
```diff
@@ -0,0 +1,32 @@
+ * This allows for a better mapping of mandatory snippets without complicating
+ * the actual indexing implementation.
+ */
+class AlternativeExpander {
```
That is an interesting point.

DailyMats commented on internal/SnippetIndex.h:
```diff
@@ -0,0 +1,46 @@
+  typedef uint32_t SnippetId;
+  typedef std::set<SnippetId> SnippetSet;
+
+  void registerSnippets(const std::string&, SnippetSet&);
```
Using `std::experimental` classes isn't very portable, and the real `std::string_view` is C++17.

Also, it does a `strlen()` (or equivalent) if the length isn't passed in the constructor. I guess that isn't a big performance penalty (especially since we always pass an `std::string`...).

DailyMats commented on internal/SnippetIndex.h:
```diff
@@ -0,0 +1,46 @@
+  void registerSnippets(const char* start, SnippetSet&);
+  void registerSnippets(const char* start, const char* end, SnippetSet&);
+
+  void getSnippets(const std::string&, SnippetSet&) const;
```
Yeah, you're right. I should trust the return value optimization.

DailyMats commented on internal/SnippetMapping.h:
```diff
@@ -0,0 +1,77 @@
+        delete e.second;
+      }
+    }
+    std::unordered_map<SnippetId, Node*> transitions_;
```
The map, which is a member of `Node`, can't have another `Node` as its value type, since the type needs to be complete (its size needs to be known, mainly):

```
/usr/include/c++/5/bits/stl_pair.h:102:11: error: ‘std::pair<_T1, _T2>::second’ has incomplete type
```

DailyMats commented on internal/SnippetMapping.h:
```diff
@@ -0,0 +1,77 @@
+                                 const Iterator& end,
+                                 const Node& node,
+                                 Result& expressions) const {
+    if (!node.expressions_.empty()) {
```
It's functionally redundant, but it has quite an impact on the performance:
```
$ time ./test do_not_check_empty
real	0m27.977s
user	0m27.968s
sys	0m0.000s

$ time ./test check_empty
real	0m1.121s
user	0m1.108s
sys	0m0.008s
```

main.cpp:
```C++
#include <unordered_map>

int main(int argc, char* argv[]) {
  bool check_empty = argc > 1 && argv[1][0] == 'c';

  std::unordered_map<int, int> m;
  m.insert(std::make_pair(1, 1234));

  std::unordered_map<int, int> m2;
  if (argc > 10000) {
    // Will not actually happen, just to avoid things getting optimized away
    m2.insert(std::make_pair(2, argc));
  }

  uint64_t n = argc > 2 ? strtoull(argv[2], 0, 10) : 1e9;
  if (check_empty) {
    for (uint64_t i = 0; i < n; i++) {
      if (!m2.empty()) {
        m.insert(m2.begin(), m2.end());
      }
    }
  } else {
    for (uint64_t i = 0; i < n; i++) {
      m.insert(m2.begin(), m2.end());
    }
  }

  return 0;
}
```

```
g++ -std=c++14 -O3 -otest main.cpp
```

cieplik replied:
Wow, thanks for sharing that.

DailyMats commented:
I've updated the branch with changes suggested in comments.

I added my own `StringView` class that works the same way as before, just wraps the start/end nicely in a simple class. I didn't rewrite it to use `RE2::Regexp` for parsing.

cieplik commented:
@DailyMats, are you going to experiment with `re2::Regexp`? The idea of reviewing the lexer/token code while potentially  being able to express our intents with higher-level abstractions scares me a bit ;-).

DailyMats commented:
@cieplik I looked at it a bit, but I don't think it's worth the trouble to rework things at this stage. I don't quite understand what scares you, the state of things right now, or the idea of more changes?

cieplik commented:
@DailyMats my primary concern is that one needs a fair amount of regexp parsing expertise to review and maintain the code. And it's easy to make a mistake or miss subtleties in such a code. `re2` had to implement all of this already, so we could reuse code that's already tested.

DailyMats commented:
@cieplik I agree that it's better to use something that is well tested and maintained than to implement our own. So I tried using `re2::Regexp`, and it's clear that it's nice to have the regexp parsed for you, but the whole logic of extracting the snippets and especially expanding the alternates is still as complicated and full of pitfalls, I would say. So I don't know if it gets all that more maintainable that way. Also, `re2::Regexp` is internal to the library and isn't exported by `make install`.

-----
Here's the core loop for the snippet indexation, for you to judge:
```C++
    void collectSnippets(re2::Regexp* e,
                         Snippets& snippets,
                         std::string& currentSnippet) {
      switch (e->op()) {
      case re2::kRegexpLiteral:
        addToSnippet(e->rune(), snippets, currentSnippet);
        break;
      case re2::kRegexpLiteralString:
        for (int i = 0; i < e->nrunes(); i++) {
          addToSnippet(e->runes()[i], snippets, currentSnippet);
        }
        break;
      case re2::kRegexpConcat:
        for (int i = 0; i < e->nsub(); i++) {
          collectSnippets(e->sub()[i], snippets, currentSnippet);
        }
        break;
      case re2::kRegexpPlus:
        collectSnippets(e->sub()[0], snippets, currentSnippet);
        collectAndResetSnippet(currentSnippet, snippets);
        break;
      case re2::kRegexpRepeat:
        if (e->min() > 0) {
          collectSnippets(e->sub()[0], snippets, currentSnippet);
          collectAndResetSnippet(currentSnippet, snippets);
        }
        break;
      case re2::kRegexpCapture:
        for (int i = 0; i < e->nsub(); i++) {
          collectSnippets(e->sub()[i], snippets, currentSnippet);
        }
        break;
      default:
        // Skip rest (and reset snippet being built)
        collectAndResetSnippet(currentSnippet, snippets);
        break;
      };
    }
```
The rest of the snippet index would still work the same way, including the loop that goes through the actual user agent string.

I didn't quite finish the alternate expansion, but it think it would require the same kind of stack that the current implementation does.

DailyMats commented:
I've recreated this repo as a private repo with the same name, and renamed the old one "uap-cpp_". 

I'm afraid I can't move the comments from the PR...

cieplik commented:
@DailyMats thanks for the example. I was hoping that using `re2` would simplify `AlternativeExpander` (and `internal/*`). But if you say it's not going to be much better then I'm fine with our current version. And you're right - we shouldn't depend on `re2`'s internals.

Adam-DM commented:
Please delete this repo once you are done with the PR review and migration thanks
